README.md
# Nagad Payment Package README

This README provides an overview of the Nagad Payment Package, its components, and usage examples. The package is designed to facilitate payment processing with the Nagad payment gateway using Python and FastAPI.

## Table of Contents
1. [Introduction](#introduction)
2. [Installation](#installation)
3. [Configuration](#configuration)
4. [Usage](#usage)
    - [Initiating a Payment](#initiating-a-payment)
    - [Handling Payment Callbacks](#handling-payment-callbacks)
    - [Verifying a Payment](#verifying-a-payment)
5. [Error Handling](#error-handling)
6. [Contributing](#contributing)
7. [License](#license)

## 1. Introduction

The Nagad Payment Package allows you to easily integrate Nagad's payment gateway into your Python applications. It provides functionality for initiating payments, handling payment callbacks, and verifying payments. This package is designed to work with the Nagad payment gateway API.

## 2. Installation

You can install the Nagad Payment Package using pip:

```bash
pip install nagadpy
```

## 3. Configuration

Before using the package, you need to configure it with your Nagad merchant information. You'll need the following information:

- `MERCHANT_ID`: Your Nagad merchant ID.
- `BASE_URL`: The base URL of the Nagad payment gateway API.
- `CALLBACK_URL`: The URL where Nagad will send payment callbacks.
- `PUBLIC_KEY`: Your Nagad public key.
- `PRIVATE_KEY`: Your Nagad private key.

You can set these values as environment variables or hardcode them in your application. Make sure to keep your private key secure.

## 4. Usage

### Initiating a Payment

To initiate a payment, you can create an instance of the `NagadPayment` class and call the `checkout_process` method. Here's an example using FastAPI:

```python
from fastapi import FastAPI
from nagadpy import NagadPayment

app = FastAPI()

# Initialize NagadPayment with your merchant information
nagad_payment = NagadPayment(
    merchant_id="YOUR_MERCHANT_ID",
    callback_url="YOUR_CALLBACK_URL",
    base_url="NAGAD_API_BASE_URL",
    public_key="YOUR_PUBLIC_KEY",
    private_key="YOUR_PRIVATE_KEY",
)

@app.get("/")
def initiate_payment():
    try:
        payment_response = nagad_payment.checkout_process(
            amount="50.00",
            invoice_number="YOUR_INVOICE_NUMBER",
        )
        # Handle the payment response, which contains the payment URL and status.
        # you can redirect url or you can send url to frontend based on status
        print(payment_response)
        #  {'callBackUrl': 'https://sandbox-ssl.mynagad.com:10061/check-out/MDkyMjIzMD', 'status': 'Success'}
    except Exception as e:
        # Handle any exceptions that may occur during the payment process
        logger.error(str(e))
    return {
        "payment_url": payment_response.get("callBackUrl"),
        "message": "success"
    }
```

### Handling Payment Callbacks

Nagad will send payment callbacks to the `CALLBACK_URL` you provided when initiating the payment. You can create an endpoint in your FastAPI application to handle these callbacks. Here's an example:

```python
from fastapi import FastAPI, Request
from nagadpy import NagadPaymentVerify

app = FastAPI()

@app.get("/payment-response")
@app.get("/payment-response", response_class=RedirectResponse, status_code=302):
    payment_response = dict(request.query_params)
    # or you can use payment_verify.parse_payment_response(query_string=params).
    # in that case you need to pass query string as str

    # handle payment_response
    order_id = payment_response.get("order_id")
    if payment_response.get("status") != "Success":
        print("Payment failed or cancel.Do the change your db")
        return f"http://127.0.0.1:8000/payment-status/?status=failed&user_tnx_ref={order_id}"

    # Verify the payment using the payment reference ID received in the callback
    payment_verify = NagadPaymentVerify(base_url="NAGAD_API_BASE_URL")
    payment_reference_id = payment_response.get("payment_ref_id")
    verification_result = payment_verify.verify_payment(payment_reference_id)

    # Handle the payment verification result
    if (
        verification_result.get("statusCode") == "000"
        and verification_result.get("status") == "Success"
        and verification_result.get("issuerPaymentRefNo")
    ):
        print("Do your business logic")
        return f"http://127.0.0.1:8000/payment-status/?status=success&user_tnx_ref={order_id}"
    else:
        print("Payment failed or cancel.Do the change your db")
        return f"http://127.0.0.1:8000/payment-status/?status=failed&user_tnx_ref={order_id}"
```

### Verifying a Payment

To verify a payment, you can use the `verify_payment` method of the `NagadPaymentVerify` class. This method sends a GET request to the Nagad payment verification endpoint. Here's an example:

```python
from nagadpy import NagadPaymentVerify

# Initialize NagadPaymentVerify with the Nagad API base URL
payment_verify = NagadPaymentVerify(base_url="NAGAD_API_BASE_URL")

# Specify the payment reference ID to verify
payment_reference_id = "YOUR_PAYMENT_REFERENCE_ID"

# Verify the payment
verification_result = payment_verify.verify_payment(payment_reference_id)

# Handle the payment verification result
print(verification_result)
```

## 5. Error Handling

The Nagad Payment Package provides custom exceptions for different stages of the payment process. You can catch and handle these exceptions to manage errors gracefully. The available exceptions include:

- `PaymentInitiationError`: Raised when an error occurs during payment initiation.
- `PaymentCompleteError`: Raised when an error occurs during payment completion.
- `PaymentVerificationError`: Raised when an error occurs during payment verification.
- `DecryptionError`: Raised when an error occurs during data decryption.
- `EncryptionError`: Raised when an error occurs during data encryption.
- `RequestError`: Raised when there is an issue with the HTTP request.
- `SignatureGenerationError`: Raised when an error occurs during signature generation.
- `CheckProcessError`: Raised when an error occurs during checkout process.

Make sure to handle these exceptions appropriately in your application.

## 6. Contributing

If you'd like to contribute to the Nagad Payment Package, please follow the standard open-source contribution guidelines, including creating issues, submitting pull requests, and adhering to the project's coding standards.

## 7. License

The Nagad Payment Package is provided under the [MIT License](LICENSE). You are free to use, modify, and distribute it according to the terms of the license.

---

This README provides a basic overview of the Nagad Payment Package. For more detailed information and examples, refer to the package's documentation and source code.


CHANGELOG.md

<a id='changelog-0.0.1'></a>
# 0.0.1 — 2023-09-29

A new scriv changelog fragment.

A new scriv changelog fragment.

## Removed

- Remove unused utils functions
- remove settings.py
- remove types

## Added

- Two utils function create to generate challenge and get current time in local timezone.

- NagadPayment class added
- Payment process methods added
- Unit test initialize

- NagadPaymentVerify class added
- verify_payment, parse_payment_response methods added
- Unit test added for this NagadPaymentVerify class
- Unit test added for this NagadPayment class

## Changed

- utils.py updated

- NagadPayment class added


tests/test_utils.py
import unittest

from nagadpy.utils import generate_challenge, get_timestamp


class TestUtils(unittest.TestCase):
    def test_generate_challenge_returns_string(self) -> None:
        result = generate_challenge(40)
        assert isinstance(result, str), "Generated challenge should be a string"

    def test_generate_challenge_has_correct_length(self) -> None:
        length = 40
        result = generate_challenge(length)
        assert len(result) == length, f"Generated challenge length should be {length}"

    def test_generate_challenge_contains_only_lowercase_letters(self) -> None:
        result = generate_challenge(40)
        assert (
            result.islower()
        ), "Generated challenge should contain only lowercase letters"

    def test_get_timestamp_returns_string(self) -> None:
        result = get_timestamp()
        assert isinstance(result, str), "Timestamp should be a string"

    def test_get_timestamp_has_correct_length(self) -> None:
        result = get_timestamp()
        assert len(result) == 14, "Timestamp length should be 14 characters"

    def test_get_timestamp_contains_only_digits(self) -> None:
        result = get_timestamp()
        assert result.isdigit(), "Timestamp should contain only digits"


tests/test_payment.py
from unittest.mock import MagicMock, patch

import pytest
from nagadpy.payment import NagadPayment


@pytest.fixture
def nagad_payment() -> NagadPayment:
    nagad_payment = NagadPayment(
        base_url="https://example.com",
        merchant_id="merchant_id",
        callback_url="https://callback.com",
        private_key="dummy_private_key",
        public_key="dummy_public_key",
    )
    return nagad_payment


def test_is_ready(nagad_payment: NagadPayment) -> None:
    assert nagad_payment.is_ready


@patch.object(NagadPayment, "_generate_signature", return_value="dummy_signature")
def test_generate_signature_method(
    mock_generate_signature: MagicMock, nagad_payment: NagadPayment
) -> None:
    data = "test_data"
    signature = nagad_payment._generate_signature(data)
    assert signature == "dummy_signature"


@patch.object(
    NagadPayment, "_encrypt_data_using_public_key", return_value="dummy_encrypted_data"
)
def test_encrypt_data_using_public_key(
    mock_encrypt_data: MagicMock, nagad_payment: NagadPayment
) -> None:
    data = "test_data"
    encrypted_data = nagad_payment._encrypt_data_using_public_key(data)
    assert encrypted_data == "dummy_encrypted_data"


@patch.object(
    NagadPayment, "_decrypt_data_using_private_key", return_value="dummy_decrypted_data"
)
def test_decrypt_data_using_private_key(
    mock_decrypt_data: MagicMock, nagad_payment: NagadPayment
) -> None:
    data = b"test_data"
    decrypted_data = nagad_payment._decrypt_data_using_private_key(data)
    assert decrypted_data == "dummy_decrypted_data"


@patch.object(NagadPayment, "_send_request", return_value={"status": "success"})
def test_send_request(
    mock_send_request: MagicMock, nagad_payment: NagadPayment
) -> None:
    data = {"key": "value"}
    url = "https://example.com/api"
    response = nagad_payment._send_request(data, url)
    assert response == {"status": "success"}


@patch.object(
    NagadPayment, "_encrypt_data_using_public_key", return_value="dummy_encrypted_data"
)
@patch.object(NagadPayment, "_generate_signature", return_value="dummy_signature")
@patch.object(NagadPayment, "_send_request", return_value={"status": "initiated"})
def test_initiate_payment(
    mock_send_request: MagicMock,
    mock_generate_signature: MagicMock,
    mock_encrypt_data: MagicMock,
    nagad_payment: NagadPayment,
) -> None:
    invoice_number = "INV12345"
    result = nagad_payment._initiate_payment(invoice_number)
    assert result == {"status": "initiated"}


@patch.object(
    NagadPayment, "_encrypt_data_using_public_key", return_value="dummy_encrypted_data"
)
@patch.object(NagadPayment, "_generate_signature", return_value="dummy_signature")
@patch.object(NagadPayment, "_send_request", return_value={"status": "completed"})
def test_complete_payment(
    mock_send_request: MagicMock,
    mock_generate_signature: MagicMock,
    mock_encrypt_data: MagicMock,
    nagad_payment: NagadPayment,
) -> None:
    amount = "100.00"
    invoice_number = "INV12345"
    challenge = "challenge123"
    payment_reference_id = "REF456"
    result = nagad_payment._complete_payment(
        amount, invoice_number, challenge, payment_reference_id
    )
    assert result == {"status": "completed"}


@patch.object(
    NagadPayment,
    "_initiate_payment",
    return_value={"sensitiveData": "dummy_sensitive_data=="},
)
@patch.object(
    NagadPayment,
    "_decrypt_data_using_private_key",
    return_value='{"paymentReferenceId": "123", "challenge": "abc"}',
)
@patch.object(NagadPayment, "_complete_payment", return_value={"status": "completed"})
def test_checkout_process(
    mock_complete_payment: MagicMock,
    mock_decrypt_data: MagicMock,
    mock_initiate_payment: MagicMock,
    nagad_payment: NagadPayment,
) -> None:
    amount = "100.00"
    invoice_number = "INV12345"
    result = nagad_payment.checkout_process(amount, invoice_number)
    assert result == {"status": "completed"}


tests/test_payment_verify.py
from unittest.mock import Mock, patch

import pytest
from nagadpy.exceptions import PaymentVerificationError
from nagadpy.payment_verify import NagadPaymentVerify


@pytest.fixture
def nagad_payment_verify() -> NagadPaymentVerify:
    return NagadPaymentVerify(base_url="https://example.com")


@patch("requests.get")
def test_verify_payment_successful(
    mock_get: Mock, nagad_payment_verify: NagadPaymentVerify
) -> None:
    response_data = {"status": "success"}
    mock_response = Mock()
    mock_response.json.return_value = response_data
    mock_get.return_value = mock_response

    payment_reference_id = "REF456789"

    result = nagad_payment_verify.verify_payment(payment_reference_id)

    assert result == response_data
    expected_url = "https://example.com/verify/payment/REF456789"
    mock_get.assert_called_once_with(expected_url)


@patch("requests.get")
def test_verify_payment_request_error(
    mock_get: Mock, nagad_payment_verify: NagadPaymentVerify
) -> None:
    mock_get.side_effect = Exception("Request error")

    payment_reference_id = "REF456789"

    with pytest.raises(PaymentVerificationError) as exc_info:
        nagad_payment_verify.verify_payment(payment_reference_id)
    assert "Request error" in str(exc_info.value)


src/nagadpy/utils.py
import datetime
import random
import socket
import string

import pytz


def generate_challenge(string_length: int) -> str:
    """
    Generate a random challenge string consisting of lowercase letters.

    Args:
        string_length (int): The length of the generated challenge string.

    Returns:
        str: A random challenge string of the specified length.
    """
    letters = string.ascii_lowercase
    return "".join(random.choice(letters) for _ in range(string_length))


def get_timestamp() -> str:
    """
    Get the current timestamp in the "YYYYMMDDHHMMSS" format for the Asia/Dhaka timezone.

    Returns:
        str: A timestamp string in the "YYYYMMDDHHMMSS" format.
    """
    tz = pytz.timezone("Asia/Dhaka")
    now = datetime.datetime.now(tz)
    return now.strftime("%Y%m%d%H%M%S")


def get_host_ip_address() -> str | None:
    """
    Get the IP address of the host machine.

    Returns:
        str or None: The IP address of the host machine or None if it cannot be determined.
    """
    try:
        hostname = socket.gethostname()
        ip_address = socket.gethostbyname(hostname)
        return ip_address
    except OSError:
        return None


example/nagad_payment.py
# example/test_nagad.py

# Import the necessary modules from the NagadPy package
from nagadpy import settings, payment

# Customize NagadPy settings using environment variables or direct assignments
settings.NAGAD_BASE_URL = "http://sandbox.mynagad.com:10080"
settings.NAGAD_MERCHANT_ID = "683002007104225"
settings.NAGAD_PUBLIC_KEY = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAjBH1pFNSSRKPuMcNxmU5jZ1x8K9LPFM4XSu11m7uCfLUSE4SEjL30w3ockFvwAcuJffCUwtSpbjr34cSTD7EFG1Jqk9Gg0fQCKvPaU54jjMJoP2toR9fGmQV7y9fz31UVxSk97AqWZZLJBT2lmv76AgpVV0k0xtb/0VIv8pd/j6TIz9SFfsTQOugHkhyRzzhvZisiKzOAAWNX8RMpG+iqQi4p9W9VrmmiCfFDmLFnMrwhncnMsvlXB8QSJCq2irrx3HG0SJJCbS5+atz+E1iqO8QaPJ05snxv82Mf4NlZ4gZK0Pq/VvJ20lSkR+0nk+s/v3BgIyle78wjZP1vWLU4wIDAQAB"
settings.NAGAD_PRIVATE_KEY = "MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQCJakyLqojWTDAVUdNJLvuXhROV+LXymqnukBrmiWwTYnJYm9r5cKHj1hYQRhU5eiy6NmFVJqJtwpxyyDSCWSoSmIQMoO2KjYyB5cDajRF45v1GmSeyiIn0hl55qM8ohJGjXQVPfXiqEB5c5REJ8Toy83gzGE3ApmLipoegnwMkewsTNDbe5xZdxN1qfKiRiCL720FtQfIwPDp9ZqbG2OQbdyZUB8I08irKJ0x/psM4SjXasglHBK5G1DX7BmwcB/PRbC0cHYy3pXDmLI8pZl1NehLzbav0Y4fP4MdnpQnfzZJdpaGVE0oI15lq+KZ0tbllNcS+/4MSwW+afvOw9bazAgMBAAECggEAIkenUsw3GKam9BqWh9I1p0Xmbeo+kYftznqai1pK4McVWW9//+wOJsU4edTR5KXK1KVOQKzDpnf/CU9SchYGPd9YScI3n/HR1HHZW2wHqM6O7na0hYA0UhDXLqhjDWuM3WEOOxdE67/bozbtujo4V4+PM8fjVaTsVDhQ60vfv9CnJJ7dLnhqcoovidOwZTHwG+pQtAwbX0ICgKSrc0elv8ZtfwlEvgIrtSiLAO1/CAf+uReUXyBCZhS4Xl7LroKZGiZ80/JE5mc67V/yImVKHBe0aZwgDHgtHh63/50/cAyuUfKyreAH0VLEwy54UCGramPQqYlIReMEbi6U4GC5AQKBgQDfDnHCH1rBvBWfkxPivl/yNKmENBkVikGWBwHNA3wVQ+xZ1Oqmjw3zuHY0xOH0GtK8l3Jy5dRL4DYlwB1qgd/Cxh0mmOv7/C3SviRk7W6FKqdpJLyaE/bqI9AmRCZBpX2PMje6Mm8QHp6+1QpPnN/SenOvoQg/WWYM1DNXUJsfMwKBgQCdtddE7A5IBvgZX2o9vTLZY/3KVuHgJm9dQNbfvtXw+IQfwssPqjrvoU6hPBWHbCZl6FCl2tRh/QfYR/N7H2PvRFfbbeWHw9+xwFP1pdgMug4cTAt4rkRJRLjEnZCNvSMVHrri+fAgpv296nOhwmY/qw5Smi9rMkRY6BoNCiEKgQKBgAaRnFQFLF0MNu7OHAXPaW/ukRdtmVeDDM9oQWtSMPNHXsx+crKY/+YvhnujWKwhphcbtqkfj5L0dWPDNpqOXJKV1wHt+vUexhKwus2mGF0flnKIPG2lLN5UU6rs0tuYDgyLhAyds5ub6zzfdUBG9Gh0ZrfDXETRUyoJjcGChC71AoGAfmSciL0SWQFU1qjUcXRvCzCK1h25WrYS7E6pppm/xia1ZOrtaLmKEEBbzvZjXqv7PhLoh3OQYJO0NM69QMCQi9JfAxnZKWx+m2tDHozyUIjQBDehve8UBRBRcCnDDwU015lQN9YNb23Fz+3VDB/LaF1D1kmBlUys3//r2OV0Q4ECgYBnpo6ZFmrHvV9IMIGjP7XIlVa1uiMCt41FVyINB9SJnamGGauW/pyENvEVh+ueuthSg37e/l0Xu0nm/XGqyKCqkAfBbL2Uj/j5FyDFrpF27PkANDo99CdqL5A4NQzZ69QRlCQ4wnNCq6GsYy2WEJyU2D+K8EBSQcwLsrI7QL7fvQ=="

settings.NAGAD_MERCHANT_CALLBACK_URL = "your_callback_url"
settings.NAGAD_INTIATE_PAYMENT_URL = "initiate_payment"
settings.NAGAD_COMPLETE_PAYMENT_URL = "complete_payment"

# Define the invoice number and host IP for testing
invoice_number = "invoice123"
host_ip = "127.0.0.1"

# Test the payment initiation
result, error = payment.nagad_checkout_process(invoice_number, "100", host_ip)

if error:
    print("Error:", error)
else:
    print("Initiate Payment Result:", result)

# # Test the complete payment
# amount = 100  # Replace with the actual amount
# challenge = "challenge123"  # Replace with the actual challenge
# payment_reference_id = "payment_reference123"  # Replace with the actual reference ID

# result, error = payment.complete_payment(
#     invoice_number, amount, challenge, payment_reference_id, host_ip
# )

# if error:
#     print("Error:", error)
# else:
#     print("Complete Payment Result:", result)


src/nagadpy/exceptions.py
class PaymentVerificationError(Exception):
    """
    Custom exception for errors encountered during payment verification.

    This exception is raised when there is an issue with verifying a payment.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class SignatureGenerationError(Exception):
    """
    Custom exception for errors encountered during signature generation.

    This exception is raised when there is an issue with generating digital signatures.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class EncryptionError(Exception):
    """
    Custom exception for errors encountered during data encryption.

    This exception is raised when there is an issue with encrypting sensitive data.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class DecryptionError(Exception):
    """
    Custom exception for errors encountered during data decryption.

    This exception is raised when there is an issue with decrypting encrypted data.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class RequestError(Exception):
    """
    Custom exception for errors encountered during HTTP requests.

    This exception is raised when there is an issue with making HTTP requests.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class PaymentInitiationError(Exception):
    """
    Custom exception for errors encountered during payment initiation.

    This exception is raised when there is an issue with initiating a payment.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class PaymentCompleteError(Exception):
    """
    Custom exception for errors encountered during payment completion.

    This exception is raised when there is an issue with completing a payment.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


class CheckProcessError(Exception):
    """
    Custom exception for errors encountered during the payment process.

    This exception is raised when there is an issue with any step of the payment process.

    Attributes:
        message (str): A human-readable error message describing the exception.
    """

    pass


src/nagadpy/payment.py
import base64
import json
from dataclasses import dataclass

import requests
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding
from nagadpy.exceptions import (
    CheckProcessError,
    DecryptionError,
    EncryptionError,
    PaymentCompleteError,
    PaymentInitiationError,
    RequestError,
    SignatureGenerationError,
)
from nagadpy.utils import generate_challenge, get_host_ip_address, get_timestamp


@dataclass
class NagadPayment:
    base_url: str
    merchant_id: str
    callback_url: str
    private_key: str
    public_key: str

    @property
    def is_ready(self) -> bool:
        """Check if all required properties are set."""
        return all(
            getattr(self, attr, None) is not None for attr in self.__annotations__
        )

    def _generate_signature(self, data: str) -> str:
        """
        Generate a digital signature for the given data using the private key.

        Args:
            data (str): The data to be signed.

        Returns:
            Tuple[str, Optional[Exception]]: A tuple containing the generated signature as a
            base64-encoded string and an optional exception if an error occurs during the
            signature generation. If the operation is successful, the second element will be None.
        """

        try:
            pk = (
                "-----BEGIN RSA PRIVATE KEY-----\n"
                + self.private_key
                + "\n-----END RSA PRIVATE KEY-----"
            )
            private_key = serialization.load_pem_private_key(
                pk.encode(), password=None, backend=default_backend()
            )
            sign = private_key.sign(data.encode(), padding.PKCS1v15(), hashes.SHA256())
            signature = base64.b64encode(sign)
            return signature.decode("utf-8")
        except Exception as e:
            raise SignatureGenerationError(
                f"Error generating signature: {str(e)}"
            ) from e

    def _encrypt_data_using_public_key(self, data: str) -> str:
        """
        Encrypt sensitive data using the provided public key.

        Args:
            data (str): The sensitive data to be encrypted.

        Returns:
            Tuple[str, Optional[Exception]]: A tuple containing the encrypted data as a
            base64-encoded string and an optional exception if an error occurs during the
            encryption. If the operation is successful, the second element will be None.
        """
        try:
            pk = (
                "-----BEGIN PUBLIC KEY-----\n"
                + self.public_key
                + "\n-----END PUBLIC KEY-----"
            )
            public_key = serialization.load_pem_public_key(
                pk.encode(), backend=default_backend()
            )
            encrypted_data = public_key.encrypt(data.encode(), padding.PKCS1v15())
            encoded_data = base64.b64encode(encrypted_data)
            return encoded_data.decode("utf-8")
        except Exception as e:
            raise EncryptionError(f"Error encrypting data: {str(e)}") from e

    def _decrypt_data_using_private_key(self, data: bytes) -> str:
        """
        Decrypts the provided encrypted data using the merchant's private key.

        Args:
            data (str): The encrypted data to be decrypted.

        Returns:
            str: The decrypted plaintext message.

        Raises:
            DecryptionError: If an error occurs during the decryption process.

        Note:
            This method assumes that the private key has been previously loaded and is stored
            in the `private_key` attribute of the class.

        Example:
            To decrypt encrypted_data using a merchant's private key, you can call this method as follows:

            decrypted_message = self._decrypt_data_using_private_key(encrypted_data)
        """

        merchant_private_key = self.private_key

        try:
            pk = (
                "-----BEGIN RSA PRIVATE KEY-----\n"
                + merchant_private_key
                + "\n-----END RSA PRIVATE KEY-----"
            )
            private_key = serialization.load_pem_private_key(
                pk.encode(), password=None, backend=default_backend()
            )
            original_message = private_key.decrypt(data, padding.PKCS1v15())
            return original_message.decode("utf-8")
        except Exception as e:
            raise DecryptionError(f"Error decrypting data: {str(e)}") from e

    def _send_request(self, data: dict, url: str) -> dict:
        """
        Sends an HTTP POST request to the specified URL with the provided JSON data and headers.

        Args:
            data (dict): A dictionary representing the JSON data to be sent in the request body.
            url (str): The URL to which the request will be sent.

        Returns:
            dict: A dictionary representing the JSON response from the server.

        Raises:
            RequestError: If there is an error during the request or if the response status code is not 200.

        Note:
            This method sends a POST request with JSON data and includes specific headers such as "Content-Type",
            "X-KM-IP-V4", "X-KM-Client-Type", and "X-KM-Api-Version".

        Example:
            To send a request with JSON data to a specific URL, you can call this method as follows:

            request_data = {"key1": "value1", "key2": "value2"}
            response_data = self._send_request(request_data, "https://example.com/api")
        """

        headers = {
            "Content-Type": "application/json",
            "X-KM-IP-V4": get_host_ip_address(),
            "X-KM-Client-Type": "PC_WEB",
            "X-KM-Api-Version": "v-0.2.0",
        }

        try:
            response = requests.post(url, json.dumps(data), headers=headers)
            json_response = response.json()
            if response.status_code != 200:
                raise RequestError(
                    f"HTTP error {response.status_code}: {json_response}"
                )
            return json_response
        except Exception as e:
            raise RequestError(f"Request error: {str(e)}")

    def _initiate_payment(self, invoice_number: str) -> dict:
        """
        Initiates a payment by sending a request to the payment gateway's initialization endpoint.

        Args:
            invoice_number (str): The unique invoice number or identifier for the payment.

        Returns:
            dict: A dictionary containing the response from the payment gateway.

        Raises:
            PaymentInitiationError: If there is an error during payment initiation, including issues with
                signature generation, data encryption, or the HTTP request.

        Example:
            To initiate a payment, you can call this method with the invoice number:

            payment_info = self._initiate_payment("INV12345")
        """

        try:
            now = get_timestamp()
            sensitive_data = {
                "merchantId": self.merchant_id,
                "orderId": invoice_number,
                "challenge": generate_challenge(40),
                "datetime": now,
            }
            sensitive_data_str = json.dumps(sensitive_data)
            encrypted_sensitive_data = self._encrypt_data_using_public_key(
                sensitive_data_str
            )
            signature = self._generate_signature(sensitive_data_str)
            data = {
                "dateTime": now,
                "sensitiveData": encrypted_sensitive_data,
                "signature": signature,
            }
            url = f"{self.base_url}/check-out/initialize/{self.merchant_id}/{invoice_number}"
            result = self._send_request(data=data, url=url)
            return result
        except (SignatureGenerationError, EncryptionError, RequestError) as e:
            raise PaymentInitiationError(str(e)) from e
        except Exception as e:
            raise PaymentInitiationError(str(e))

    def _complete_payment(
        self,
        amount: str,
        invoice_number: str,
        challenge: str,
        payment_reference_id: str,
    ) -> dict:
        """
        Completes a payment by sending a request to the payment gateway's completion endpoint.

        Args:
            amount (str): The amount to be paid.
            invoice_number (str): The unique invoice number or identifier for the payment.
            challenge (str): A challenge associated with the payment.
            payment_reference_id (str): The reference identifier for the payment.

        Returns:
            dict: A dictionary containing the response from the payment gateway.

        Raises:
            PaymentCompleteError: If there is an error during payment completion, including issues with
                signature generation, data encryption, or the HTTP request.

        Example:
            To complete a payment, you can call this method with the required parameters:

            payment_info = self._complete_payment("100.00", "INV12345", "challenge123", "REF456")
        """

        try:
            sensitive_data = {
                "merchantId": self.merchant_id,
                "orderId": invoice_number,
                "currencyCode": "050",
                "amount": amount,
                "challenge": challenge,
            }

            sensitive_data_str = json.dumps(sensitive_data)
            encrypt_result = self._encrypt_data_using_public_key(sensitive_data_str)

            signature_result = self._generate_signature(sensitive_data_str)

            data = {
                "dateTime": get_timestamp(),
                "sensitiveData": encrypt_result,
                "signature": signature_result,
                "merchantCallbackURL": self.callback_url,
                "additionalMerchantInfo": {},
            }

            url = f"{self.base_url}/check-out/complete/{payment_reference_id}"

            result = self._send_request(data=data, url=url)
            return result
        except (SignatureGenerationError, EncryptionError, RequestError) as e:
            raise PaymentCompleteError(str(e)) from e
        except Exception as e:
            raise PaymentCompleteError(str(e))

    def checkout_process(self, amount: str, invoice_number: str) -> dict:
        """
        Initiates and completes the payment process for a given amount and invoice number.

        Args:
            amount (str): The amount to be paid.
            invoice_number (str): The unique invoice number or identifier for the payment.

        Returns:
            dict: A dictionary containing the response from the payment gateway after completion.
            Example:  ```{"callBackUrl": "", "status": ""}```

        Raises:
            PaymentCompleteError: If there is an error during the payment process, including issues with
                initiation, decryption, completion, or missing required data.


        Example:
            To perform a payment process, you can call this method with the amount and invoice number:
            ```
            nagad_payment = NagadPayment(
                merchant_id=MERCHANT_ID,
                callback_url=CALLBACK_URL,
                base_url=BASE_URL,
                public_key=PUBLIC_KEY,
                private_key=PRIVATE_KEY,
            )
            payment_response = nagad_payment.checkout_process("100.00", "INV12345")
            ```

        Callback Response:
            Method: GET
            QueryParams Example:
            ```
            {
                "merchant": "",
                "order_id": "",
                "payment_ref_id": "",
                "status": "",
                "status_code": "",
                "message": ""
            }
            ```
        """
        if not self.is_ready:
            raise ValueError("All payment information properties must be set")
        try:
            initiated_result = self._initiate_payment(invoice_number=invoice_number)

            sensitive_data = initiated_result.get("sensitiveData")

            if not sensitive_data:
                raise CheckProcessError(
                    "sensitive data is missing from initiated_result"
                )

            decrypted_result = self._decrypt_data_using_private_key(
                base64.b64decode(sensitive_data)
            )

            decrypted_sensitive_data_dict = json.loads(decrypted_result)

            if not decrypted_sensitive_data_dict.get(
                "paymentReferenceId"
            ) or not decrypted_sensitive_data_dict.get("challenge"):
                raise CheckProcessError(
                    "paymentReferenceId and challenge data is missing from initiated_result"
                )

            payment_reference_id = decrypted_sensitive_data_dict.get(
                "paymentReferenceId"
            )
            challenge = decrypted_sensitive_data_dict.get("challenge")
            complete_result = self._complete_payment(
                amount=amount,
                invoice_number=invoice_number,
                challenge=challenge,
                payment_reference_id=payment_reference_id,
            )
            return complete_result
        except (DecryptionError, PaymentInitiationError, PaymentCompleteError) as e:
            raise PaymentCompleteError(str(e)) from e
        except Exception as e:
            raise PaymentCompleteError(str(e))
